<a href="https://colab.research.google.com/github/dondreojordan/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2/Jordan_Dondre_Group_TL_Maggie_DS17_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
#Cannot do comparison with the following:
# sex and internet
# failures and study time

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [0]:
from scipy import stats
import pandas as pd
import numpy as np



In [0]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-06-09 19:35:40--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.06s   

2020-06-09 19:35:41 (321 KB/s) - ‘student.zip’ saved [20478/20478]



In [0]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [0]:
df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 

**Null Hypothesis: These two variables are "independent" there's no relationship between the two**

**Alternative Hypothesis: The two variables are "dependent" there is a relationship or association between them.**


>>Use Chi^2 tests and stat.chi2contingency to find Two (2) pairs of variables that are **dependent** (are associated with one another).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 

In [0]:
#1st Dependent Example: Relationship of 'age' and 'failures'

print(df['age'].value_counts().sort_index())
print(df['failures'].value_counts().sort_index())

15     82
16    104
17     98
18     82
19     24
20      3
21      1
22      1
Name: age, dtype: int64
0    312
1     50
2     17
3     16
Name: failures, dtype: int64


In [0]:
contingencydep1 = pd.crosstab(df['age'], df['failures'])

contingencydep1

failures   0   1  2  3
age                   
15        71   3  5  3
16        88  10  4  2
17        84   8  2  4
18        63  16  1  2
19         5  13  3  3
20         1   0  2  0
21         0   0  0  1
22         0   0  0  1

*Null Hypothesis*: The variables 'age' and 'failures'are "independent" there's no relationship between the two

>>OR

*Alternative Hypothesis*: The variables 'age' and 'failures' are "dependent" there is a relationship or association between them.

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingencydep1)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 144.29903924777085
p value 2.1137084383846074e-20
degrees of freedom 21
expected frequencies table [[6.47696203e+01 1.03797468e+01 3.52911392e+00 3.32151899e+00]
 [8.21468354e+01 1.31645570e+01 4.47594937e+00 4.21265823e+00]
 [7.74075949e+01 1.24050633e+01 4.21772152e+00 3.96962025e+00]
 [6.47696203e+01 1.03797468e+01 3.52911392e+00 3.32151899e+00]
 [1.89569620e+01 3.03797468e+00 1.03291139e+00 9.72151899e-01]
 [2.36962025e+00 3.79746835e-01 1.29113924e-01 1.21518987e-01]
 [7.89873418e-01 1.26582278e-01 4.30379747e-02 4.05063291e-02]
 [7.89873418e-01 1.26582278e-01 4.30379747e-02 4.05063291e-02]]


>>Based on a chi^2 statistic of 144.299 and a p-value of 2.113e-20, I reject the null hypothesis, that observed frequencies are equal to expected frequencies and suggest the alternative that the two variables are "dependent". There is a relationship associated between them.

In [0]:
# 2nd Dependent Example: Relationship of 'nursery' and 'activities'

print(df['nursery'].value_counts().sort_index())
print(df['activities'].value_counts().sort_index())

no      81
yes    314
Name: nursery, dtype: int64
no     194
yes    201
Name: activities, dtype: int64


In [0]:
contingencydep2 = pd.crosstab(df['nursery'], df['activities'])

contingencydep2

activities   no  yes
nursery             
no           40   41
yes         154  160

*Null Hypothesis*: The variables 'nursery' and 'activities'are "independent" there's no relationship between the two

>>OR

*Alternative Hypothesis*: The variables 'nursery' and 'failures' are "dependent" there is a relationship or association between them.

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingencydep2)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 0.004951472553918086
p value 0.9439017945961875
degrees of freedom 1
expected frequencies table [[ 39.78227848  41.21772152]
 [154.21772152 159.78227848]]


Based on a chi^2 statistic of .004 and a p-value of .943, I reject the null hypothesis, that observed frequencies are equal to expected frequencies (that this is a fair die) and suggest the alternative, that this is an unfair die.

>>Use Chi^2 tests and stat.chi2contingency to find Two (2) pairs of variables that are **independent** (are associated with one another).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another?

In [0]:
#1st Independent Example: Relationship of 'sex' and 'romantic'.

print(df['sex'].value_counts().sort_index())
print(df['romantic'].value_counts().sort_index())

F    208
M    187
Name: sex, dtype: int64
no     263
yes    132
Name: romantic, dtype: int64


In [0]:
contingencyind1 = pd.crosstab(df['sex'], df['romantic'])

contingencyind1

romantic   no  yes
sex               
F         129   79
M         134   53

*Null Hypothesis*: The variables 'sex' and 'romantic' are "independent" there's no relationship between the two

>>OR

*Alternative Hypothesis*: The variables 'sex' and 'romantic' are "dependent" there is a relationship or association between them.

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingencyind1)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 3.689657998622032
p value 0.05475084154682915
degrees of freedom 1
expected frequencies table [[138.49113924  69.50886076]
 [124.50886076  62.49113924]]


>>Conclusion: There is not enough evidence to conclude that sex and romance are associated **(show a relationship)**. They might be independent.

In [0]:
# 2nd Independent Example: Relationship of 'school' and 'absences'.

print(df['school'].value_counts().sort_index())
print(df['absences'].value_counts().sort_index())

In [0]:
contingencyind2 = pd.crosstab(df['school'], df['absences'])

contingencyind2

absences  0   1   2   3   4   5   6   7   ...  26  28  30  38  40  54  56  75
school                                    ...                                
GP        99   3  59   5  44   3  31   5  ...   1   1   1   1   1   1   1   1
MS        16   0   6   3   9   2   0   2  ...   0   0   0   0   0   0   0   0

[2 rows x 34 columns]

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingencyind2)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 34.321617446107915
p value 0.4042095270470631
degrees of freedom 33
expected frequencies table [[101.60759494   2.65063291  57.43037975   7.06835443  46.8278481
    4.41772152  27.38987342   6.18481013  19.43797468   2.65063291
   15.02025316   2.65063291  10.60253165   2.65063291  10.60253165
    2.65063291   6.18481013   0.8835443    4.41772152   0.8835443
    3.53417722   0.8835443    2.65063291   0.8835443    0.8835443
    0.8835443    0.8835443    0.8835443    0.8835443    0.8835443
    0.8835443    0.8835443    0.8835443    0.8835443 ]
 [ 13.39240506   0.34936709   7.56962025   0.93164557   6.1721519
    0.58227848   3.61012658   0.81518987   2.56202532   0.34936709
    1.97974684   0.34936709   1.39746835   0.34936709   1.39746835
    0.34936709   0.81518987   0.1164557    0.58227848   0.1164557
    0.46582278   0.1164557    0.34936709   0.1164557    0.1164557
    0.1164557    0.1164557    0.1164557    0.1164557    0.1164557
    0.1164557    0.1164557    0.1164557

>>Conclusion: There is not enough evidence to conclude that the school and number of absneces are associated **(show a relationship)**. They might be independent.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

# **EXAMPLE 1**

In [0]:
#1st Independent Example: Relationship of 'sex' and 'romantic'.

print(df['sex'].value_counts().sort_index())
print(df['romantic'].value_counts().sort_index())

F    208
M    187
Name: sex, dtype: int64
no     263
yes    132
Name: romantic, dtype: int64


In [0]:
observedind1 = pd.crosstab(df['sex'], df['romantic'])

observedind1

romantic   no  yes
sex               
F         129   79
M         134   53

In [0]:
observedind1_with_margins = pd.crosstab(df['sex'], df['romantic'], margins=True)

observedind1_with_margins

romantic   no  yes  All
sex                    
F         129   79  208
M         134   53  187
All       263  132  395

In [0]:
observedind1 = observedind1.values # Make an expected value table that matches the size and shape of our observed value table
observedind1

array([[129,  79],
       [134,  53]])

In [0]:
observedind1[1][1] 

53

In [0]:
# (row_total)(column_total) / sample_size 

expectedind1_row1_col1 = 208*263/395


expectedind1_row1_col1


138.49113924050633

In [0]:
row_totals = df['sex'].value_counts().values # row_totals contains 'sex'

row_totals

array([208, 187])

In [0]:
col_totals = df['romantic'].value_counts().values # col_totals contains 'romance'

col_totals

array([263, 132])

In [0]:
sample_size = df.shape[0] # sample size

sample_size

395

In [0]:
expectedind1 = np.array([[0,0],
                     [0,0]])

for i, col in enumerate(col_totals):
  for j, row in enumerate(row_totals): 
    expected_value = (row*col / sample_size)
    expectedind1[i][j] = expected_value


expectedind1

array([[138, 124],
       [ 69,  62]])

In [0]:
observedind1

array([[129,  79],
       [134,  53]])

Calculate the chi^2 statistic (test statistic)
χ2=∑(observedi−expectedi)2(expectedi)

In [0]:
((observedind1 - expected)**2 / expected).sum().sum()

79.45593735390369

# **EXAMPLE 2**

In [0]:

#1st Dependent Example: Relationship of 'age' and 'failures'

print(df['age'].value_counts())
print(df['failures'].value_counts())

16    104
17     98
18     82
15     82
19     24
20      3
22      1
21      1
Name: age, dtype: int64
0    312
1     50
2     17
3     16
Name: failures, dtype: int64


In [0]:
observeddep1 = pd.crosstab(df['age'], df['failures'])

observeddep1

failures   0   1  2  3
age                   
15        71   3  5  3
16        88  10  4  2
17        84   8  2  4
18        63  16  1  2
19         5  13  3  3
20         1   0  2  0
21         0   0  0  1
22         0   0  0  1

## Expected Value Calculation
\begin{align}
expected_{i,j} =\frac{(row_{i} \text{total})(column_{j} \text{total}) }{(\text{total observations})}  
\end{align}

In [0]:
observeddep1_with_margins = pd.crosstab(df['age'], df['failures'], margins=True)

observeddep1_with_margins

failures    0   1   2   3  All
age                           
15         71   3   5   3   82
16         88  10   4   2  104
17         84   8   2   4   98
18         63  16   1   2   82
19          5  13   3   3   24
20          1   0   2   0    3
21          0   0   0   1    1
22          0   0   0   1    1
All       312  50  17  16  395

In [0]:
# Make an expected value table that matches the size and shape of our observed value table

In [0]:
observeddep1 = observed.values
observeddep1

NameError: ignored

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [0]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

>>The sample mean is an "estimate" for the population mean

- What are hypothesis test? How are they useful?

>>Hypothesis tests are used to show if the is a difference between the means in two unrelated groups. Most of the time, we are looking to reject the 'standard', or null, hypothesis that says, in short, the data is independent of one another and the mean populations are not equal.

- What is a "null hypothesis?"

>>The Null Hypothesis says "for the independent t-test, the populations means from the two unrelated groups are equal." This is the "boring" hypothesis proving that you accept that the population means of two unrelated groups are equal.

>>To suggest the alternative hypothesis is to accept that the population means are not equal. 

- What is a p-value and what does it represent? 

>A smaller p-value means that there is stronger evidence in favor of the alternative hypothesis. The pvalue is represent if p value is less than the threshold, you reject the null hypothesis. if its greater than the threshold, you fail to reject.

- What does it mean for something to be "statistically significant?"

"Statistically statistic" results the the Null Hypothesis is true and that there isn't really any difference

- What is a test statistic and how does it relate to a p-value?

>>

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

>>**Degrees of Freedom in statistics refer to number of values in a calulation that have the "freedom" to vary.**

For example: The degrees of freedom for a set of three numbers is TWO.

>>**For a 1-sample t-test, you take the sample size (n) and the calculation (n - 1) to get the "degrees of freedom" since the last value in the sample size is constrained from variance. 

>>For a 2-sample, t-test, the degrees of freedom compares 

>>For a Chi-Square Test of Independence, the degrees of freedon are the number of cells in tabular data that can vary, not included fixed row and columnof margin_headers/margin_totals. This test is used to compare two categorical variable against one another for dependency. 

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)